File: Level_Order_Traversal_DFS.py
# Created by Aashish Adhikari at 6:13 PM 1/20/2021

'''
Time Complexity:
O(n)

Space Complexity:
Ignoring recursive stack, O(1)
If not, O(h) since we store a maximum of h items in the recursive stack at any given time.


'''

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def helper(self, node, level):



        # logic
        if level+1 > len(self.sol):
            self.sol.append([])

        self.sol[level].append(node.val)

        if node.left is not None:
            self.helper(node.left, level+1)

        if node.right is not None:
            self.helper(node.right, level+1)


    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []

        self.sol = []

        self.helper(root, 0)

        return self.sol

File: level-order-traversal_dfs.java
// Time Complexity : O(n)
// Space Complexity : O(h) .. O(n) for skewed tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : 

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root==null) return new ArrayList<>();
        
        /* DFS */
        
        List<List<Integer>> result=new ArrayList<>();
        dfs(root,0,result);
    
        return result;
    }
    
    private void dfs(TreeNode root, int level, List<List<Integer>> result)
    {
        /* Base Case */
        if(root==null)
            return;
        
        /* Action */
        
        if(level==result.size())
            result.add(new ArrayList<>());
        
        result.get(level).add(root.val);
        /* Recurse */
        
        dfs(root.left,level+1,result);
        dfs(root.right,level+1,result);
    }
}